### ISSUE TITLE
Consider employing multi-valued logic within the relation checker

### ISSUE THREAD

**Issue by mtshiba** [astral-sh/ty#4050]
As described in astral-sh/ruff#26503, subtyping problems in the Python type system are generally undecidable, so using bool as the return value of `is_subtype_of/is_redundant_with` is incomplete. This will create motivation to adopt three-valued logic (Kleene K3), which adds `Indeterminate` to `{True, False}`, which represents "neither the truth nor falsehood could be proven".

A direct example of the need to adopt this is union reduction of recursive aliases (working on a fix at https://github.com/astral-sh/ruff/pull/26882). Complete reduction of unions involving recursive aliases is undecidable because the general subtyping(redundancy) problem is undecidable. Neither True/False here is correct as a conservative fallback. If `True`, there is a possibility of accidentally removing element types that should remain. If `False`, the recursive type will remain, and infinite expansion may not be avoided. If `Indeterminate` can be used as a judgment, it is possible to prevent excessive removal while avoiding infinite expansion by adding `Unknown` to the reduction result.

---

To generalize further, we may need to adopt four-valued logic (Belnap-Dunn logic, FDE), which includes `Inconsistent`​ in addition to `Indeterminate`. A data type representing four-valued logic could be defined as follows:

```rust
pub(crate) struct RelationConstraintSet<'db> { 
    positive_evidence: ConstraintSet<'db>, 
    negative_evidence: ConstraintSet<'db>,
}
```

```
(p, n): ({Always, Never}, {Always, Never})

True = (Always, Never)
False = (Never, Always)
Indeterminate = (Never, Never)
Inconsistent = (Always, Always)
```

`Inconsistent` represents "some evidence is true and some evidence is false". This may be useful when users specify inconsistent typings. For example:

```python
class Base: 
    def foo(self, x: int) -> None: ...

class Child(Base): 
    def foo(self, x: str) -> str: ... # invalid-method-override

class HasStringFoo(Protocol): 
    def foo(self, x: str) -> str: ...

# currently: Intersection[Base, HasStringFoo] = Never
def use(x: Intersection[Base, HasStringFoo]) -> None: ...

use(Child()) # invalid-argument-type
```

The code is inconsistent in the sense that the user is overriding the method incorrectly, and `use(Child())` seems to have to be an error. However, since it is `Child <: Base`, it may not be an error.
So the apparent invalid override actually has a two-way basis.

* positive evidence: `Child <: Base` due to nominal inheritance
* negative evidence: actual member signature violates `Base` contract

With contradiction-tolerant logic (∋ FDE), it would be possible to diagnose invalid-method-override while silencing invalid-argument-type. In this system, `Intersection[Base, HasStringFoo] ≠ Never`.

Alternatively, `Inconsistent` ​​may be useful for expressing logical inconsistencies inherent in the Python type system. For example, `Hashable` violates the Liskov Substitution Principle (c.f. https://github.com/astral-sh/ty/issues/1132).

---

Note that 3-valued logic and 4-valued logic are "weak" logics, so some of the theorems that we had previously assumed to hold can no longer be used. They may break some type property tests. The weaker the logic, the more it will require a rethinking of the nature of the type system. If K3 is adopted, the law of excluded middle (contradiction law) will no longer hold. Furthermore, if FDE is adopted, explosion rules, disjunctive syllogisms, etc. will no longer hold.

I don't think K3 will radically worsen the nature of the type system, but FDE explicitly allows for contradictions, so there is a possibility that the nature of the type system will become noticeably worse. However, in reality, corruption only occurs when the type definition itself is invalid, so applying a filter such as "Is this a well-defined type?" will recover the properties.

**Comment by carljm** [comment]
I do not think we should consider adding four-valued logic; in general I think that given inconsistent inputs, inconsistent results can be expected. In the given example it is fine if we consider that intersection to be `Never` and don't allow `Child` to be assigned to it.

There are known issues with potential non-determinism due to union/intersection simplification, structural types, and Liskov-violating types -- I think we need to do something here to ensure determinism, but I think that can be addressed without four-valued logic.

Regarding three-valued logic, this seems potentially more interesting. I think losing the law of excluded middle is quite problematic if it requires that we no longer treat `P & ~P` as `Never`; this is quite important for type narrowing via intersections to work as expected.

Ultimately we have to give a yes/no answer to assignability questions -- how would an answer of "indeterminate" function?

I think it is acceptable if sometimes we give wrong answers in some recursive cases -- it's more a practical question of giving sufficiently-correct answers in the commonly encountered cases.

**Comment by carljm** [comment]
This may not really be relevant to the core point here, but note also #3618 means that technically `Base & HasStringFoo` are not disjoint even if we exclude Liskov violations, since a valid subclass of `Base` could define an overloaded `foo` method that is both a valid override of `Base.foo` and a valid implementation of `HasStringFoo`.

**Comment by mtshiba** [comment]
The fact that the subtyping problem is undecidable does *not* mean that "each instance of the problem is true or false, but there is no general algorithm that can solve it", but rather that "there are propositions within the system that cannot be proved to be true or false in principle". This is an appearance of Gödel's incompleteness theorems on the type system. It is unavoidable to assign `Indeterminate` to such propositions. Perhaps such an answer should be called `Independent`, as distinct from `Indeterminate`, which can be determined to be true or false within the system (but for which the type checker could not find a proof in finite time). However, the type checker will never be able to distinguish between `Indeterminate` and `Independent`.

This happens because Python's type system is Turing-complete[^1]. Due to the claim of the paper, it is possible to realize a pair of types such that `A <: B` if and only if a certain Turing machine `M` stops for a certain input `w`. Our assumption is that `A <: B` or `not A <: B` is provable within a type system. If `A <: B` is provable, it can be proven that `M(w)` stops. If `not A <: B` can be proven, it can be proven that `M(w)` does not stop. This contradicts the undecidability of the halting problem, so it is denied that `A <: B` or `not A <: B` is always provable in a type system.

So, what can we say about types whose relationship is determined to be indeterminate? Nothing.
The diagnostic that is triggered when A is a subclass of B is not triggered in the indeterminate case. The diagnostic that is triggered when A is not a subclass of B is also not triggered in the indeterminate case. This is the concrete meaning that the law of excluded middle does not hold.

[^1]: https://arxiv.org/abs/2208.14755

**Comment by mtshiba** [comment]
Even if K3 is adopted, the reduction rules regarding union and intersection will be maintained as before. This is because most reduction rules are of the form that apply when `A <: B`, and there is little that can be said about the case `not A <: B`. K3 expands the answer of `is_subtype_of(A, B)` to `{True, False, Indeterminate}`. If it is `True`, simplification rules can be applied as in the case of bool. When `False, Indeterminate`, nothing can be done, but this only adds cases where nothing can be done.

| is_subtype_of(A, B) | A \| B | A & B |
| ----- | --- | --- |
| True | B | A |
| False | A \| B | A & B |
| Indeterminate | A \| B | A & B |

Specifically, for fully static types, the following holds on K3 as well.

```
object | T = object
object & T = T
Never | T = T
Never & T = Never
T | ~T = object
T & ~T = Never
~~T = T
De Morgan rules
associativity
commutativity
absorption
...
```

Therefore, as far as reduction rules are concerned, `Indeteminate` is indistinguishable from `False`.

`Indeterminate` makes a difference when neither of the two choices can be chosen, or when the negation of a judgment is involved.

**Comment by carljm** [comment]
> The fact that the subtyping problem is undecidable does _not_ mean that "each instance of the problem is true or false, but there is no general algorithm that can solve it", but rather that "there are propositions within the system that cannot be proved to be true or false in principle". This is an appearance of Gödel's incompleteness theorems on the type system. It is unavoidable to assign `Indeterminate` to such propositions.

Yes, I understand this. What I meant by "we have to give a yes/no answer to assignability questions" is simply that ultimately we must either treat "indeterminate" as a yes or as a no, for whatever purpose we asked the question (emitting a diagnostic, simplifying a union/intersection, etc) -- we cannot give up and tell the user that their program is statically undecidable 😆  But your explanation of how we should treat it in each case seems reasonable, as does your explanation of what it actually means to lose the law of excluded middle. This all seems feasible to me. Can you give more specific examples of cases in which a result of `Indeterminate` will help us to give a more correct answer than an earlier fallback to `true` or `false` allows?

**Comment by mtshiba** [comment]
How to handle `Indeterminate` is clear based on the four-valued logic explained earlier. Here, logical values ​​are classified by "Is there positive/negative evidence?"

| Logical value | Is there positive evidence? | Is there negative evidence? |
| ------------- | --------------------------- | --------------------------- |
| True          | Yes                         | No                          |
| False         | No                          | Yes                         |
| Indeterminate | No                          | No                          |
| Inconsistent  | Yes                         | Yes                         |

In other words, the judgment we have been using up until now, "Is it foo?", can be rephrased as "Is there positive evidence that it is foo?"
When the judgment is true, it can be assumed that "it is foo" in that if block (unless `Inconsistent` is accepted). Conversely, we can't do much with an else block (don't assume "it isn't foo").
Instead of assuming "it isn't foo" in the else block, the if-else statement should be replaced with a match statement. Since K3 is an extension of bool, propositions that are True or False in K3 also hold true in bool (the reverse is not true). In addition, we only need to add handling of the Indeterminate case.

**Comment by mtshiba** [comment]
If there is a possibility of `Indeterminate` as a relation judgment result, there will be a difference in results such as narrowing. This is a fact that has already been verified through experimental implementation.

For example, consider the following recursive alias:

```python
type Shift[T1, T2, T3] = T1 | Shift[T2, T3, T1]
type Recursive[T, U] = Shift[T, T, U]
```

Since our relation checker restricts the occurrence of recursive aliases in different specializations, it is not possible to perform complete reduction and type relation checks on such types. In the example above, we know that `Recursive[T, U]` actually saturates to `T | U`, but the computation can be easily exploded by increasing the number of type variables, and it is also possible to construct an example that is undecidable in the first place.

In the world of K3, the fact that "if there is no overlap, it is disjoint" breaks down. Disjointness should be understood as a ternary value of `{Disjoint, Overlap, Indeterminate}`.
If we only need a one-way positive relation check, `False` can essentially perform the same role as `Indeterminate`. However, if we require bidirectional assignability for fields such as mutable TypedDict/Protocol, if-else narrowing will not give correct results unless the cycle fallback is `Indeterminate`.

Consider the following code:

```python
from typing import Never, TypedDict, reveal_type
from typing_extensions import TypeIs

type IntOrStr = Recursive[int, str]

class IntOrStrField(TypedDict): 
    item: IntOrStr

class IntField(TypedDict): 
    item: int

def is_int_field(value: object) -> TypeIs[IntField]: 
    return isinstance(value, dict) and isinstance(value.get("item"), int)

def check(value: IntOrStrField): 
    if is_int_field(value): 
        reveal_type(value) # IntOrStrField & IntField 
    else: 
        reveal_type(value) # IntOrStrField & ~IntField
```

Since we cannot obtain positive subtyping evidence or disjointness evidence for recursive aliases, any result where the narrowed result is `Never`, whether in the if branch or the else branch, is incorrect here.

```
False fallback: 
    is_disjoint_from(IntOrStrField, IntField) == True 
    => if: Never, else: IntOrStrField
True fallback: 
    is_assignable_to(IntOrStrField, IntField) == True 
    => if: IntOrStrField, else: Never
Indeterminate fallback: 
    No evidence for subtype or disjoint 
    => if: IntOrStrField & IntField, else: IntOrStrField & ~IntField
```
